src/raftstore/store/fsm/store.rs

```diff
@@ -508,50 +840,169 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+                "[region {}] tombstone peer [epoch: {:?}] \
+                 receive a stale message {:?}",
+                region_id, region_epoch, msg_type,
             );
-            if !self.handle_destroy_peer(job) {
-                return Ok(false);
-            }
-            has_peer = false;
+
+            let not_exist = util::find_peer(region, from_store_id).is_none();
+            Self::handle_stale_msg(
+                trans,
+                msg,
+                region_epoch,
+                is_vote_msg && not_exist,
+                None,
+                raft_metrics,
+            );
+
+            return Ok(true);
+        }
+
+        if from_epoch.get_conf_ver() == region_epoch.get_conf_ver() {
+            raft_metrics.message_dropped.region_tombstone_peer += 1;
+            return Err(box_err!(
+                "tombstone peer [epoch: {:?}] receive an invalid \
+                 message {:?}, ignore it",
+                region_epoch,
+                msg_type
+            ));
+        }
+
+        Ok(false)
+    }
+    pub fn on_raft_message(&mut self, ctx: &mut PollContext<T, C>, msg: RaftMessage) -> Result<()> {
+        let region_id = msg.get_region_id();
+        match ctx.router.send(region_id, PeerMsg::RaftMessage(msg)) {
+            Ok(()) | Err(TrySendError::Full(_)) => return Ok(()),
+            Err(TrySendError::Disconnected(PeerMsg::RaftMessage(m))) => msg = m,
+            e => panic!(
+                "{} [region {}] unexpected redirect error: {:?}",
+                self.tag, region_id, e
+            ),
         }
 
-        if has_peer {
+        debug!(
+            "{} [region {}] handle raft message {:?}, from {} to {}",
+            self.tag,
+            region_id,
+            msg.get_message().get_msg_type(),
+            msg.get_from_peer().get_id(),
+            msg.get_to_peer().get_id()
+        );
+
+        if msg.get_to_peer().get_store_id() != self.store_id() {
+            warn!(
+                "[region {}] store not match, to store id {}, mine {}, ignore it",
+                region_id,
+                msg.get_to_peer().get_store_id(),
+                self.store_id()
+            );
+            ctx.raft_metrics.message_dropped.mismatch_store_id += 1;
+            return Ok(());
+        }
+
+        if !msg.has_reigon_epoch() {
+            error!(
+                "[region {}] missing epoch in raft message, ignore it",
+                region_id
+            );
+            ctx.raft_metrics.message_dropped.mismatch_region_epoch += 1;
+            return Ok(());
+        }
+        if msg.get_is_tombstone() || msg.has_merge_target() {
+            // Target tombstone peer doesn't exist, so ignore it.
+            return;
+        }
+        if self.check_msg(&msg)? {
+            return Ok(());
+        }
+        if !self.maybe_create_peer(region_id, msg)? {
+            return Ok(());
+        }
+        let _ = ctx.router.send(region_id, PeerMsg::RaftMessage(msg));
```

Comment:
seems that it can be moved into above if-block of `maybe_create_peer`

Reply:
No, it makes sense only when target peer is created.